gdal/port/cpl_swift.cpp

```diff
@@ -170,6 +184,258 @@ bool VSISwiftHandleHelper::GetConfiguration(CPLString& osStorageURL,
+        {
+            CPLJSONObject projectDomain;
+            projectDomain.Add("name", osProjectDomainName);
+            project.Add("domain", projectDomain);
+        }
+
+        scope.Add("project", project);
+    }
+
+    CPLJSONObject auth;
+    auth.Add("identity", identity);
+    if( ! scope.GetChildren().empty() )
+        auth.Add("scope", scope);
+
+    CPLJSONObject obj;
+    obj.Add("auth", auth);
+    return obj;
+}
+
+/************************************************************************/
+/*                      GetAuthV3StorageURL()                           */
+/************************************************************************/
+
+bool VSISwiftHandleHelper::GetAuthV3StorageURL(const CPLHTTPResult *psResult,
+                                               CPLString& storageURL)
+{
+    if( psResult->pabyData == nullptr )
+        return false;
+
+    CPLJSONDocument resultJson;
+    resultJson.LoadMemory(psResult->pabyData);
+    CPLJSONObject result(resultJson.GetRoot());
+
+    CPLJSONObject token(result.GetObj("token"));
+    if( ! token.IsValid() )
+        return false;
+
+    CPLJSONArray catalog(token.GetArray("catalog"));
+    if( ! catalog.IsValid() )
+        return false;
+
+    CPLJSONArray endpoints;
+    for(int i = 0; i < catalog.Size(); ++i)
+    {
+        CPLJSONObject item(catalog[i]);
+        if( item.GetString("name") == "swift" )
+        {
+            endpoints = item.GetArray("endpoints");
+            break;
+        }
+    }
+
+    if( endpoints.Size() == 0 )
+        return false;
+
+    CPLString osRegionName = CPLGetConfigOption("SWIFT_REGION_NAME", "");
+    if( osRegionName.empty() )
+    {
+        CPLJSONObject endpoint(endpoints[0]);
+        storageURL = endpoint.GetString("url");
+        return true;
+    }
+
+    for(int i = 0; i < endpoints.Size(); ++i)
+    {
+        CPLJSONObject endpoint(endpoints[i]);
+        if( endpoint.GetString("region") == osRegionName )
+        {
+            storageURL = endpoint.GetString("url");
+            CPLDebug("SWIFT", "Storage URL '%s' for region '%s'", storageURL.c_str(), osRegionName.c_str());
+            return true;
+        }
+    }
+
+    return false;
+}
+
+/************************************************************************/
+/*                                AuthV3()                              */
+/************************************************************************/
+
+bool VSISwiftHandleHelper::AuthV3(CPLString& osStorageURL,
+                                  CPLString& osAuthToken)
+{
+    CPLString osUser = CPLGetConfigOption("SWIFT_USER", "");
+    CPLString osKey = CPLGetConfigOption("SWIFT_KEY", "");
+    CPLJSONObject postObject(CreateAuthV3RequestObject());
+    std::string post = postObject.Format(CPLJSONObject::PrettyFormat::Plain);
+
+    CPLString osAuthURL = CPLGetConfigOption("SWIFT_AUTH_V3_URL", "");
+    std::string url = osAuthURL;
+    if( url.back() != '/' )
```

Comment:
```suggestion
    if( !url.empty() && url.back() != '/' )
```

Reply:
Fixed to be more foolproof, but it's not absolutely necessary as the code will never get there if `SWIFT_AUTH_V3_URL` is empty or unset.